lib/services/role.go

```diff
@@ -505,6 +505,33 @@ func ApplyTraits(r Role, traits map[string][]string) Role {
 		if inLabels != nil {
 			r.SetDatabaseLabels(condition, applyLabelsTraits(inLabels, traits))
 		}
+
+		// apply templates to impersonation conditions
+		inCond := r.GetImpersonateConditions(condition)
+		var outCond types.ImpersonateConditions
+		for _, user := range inCond.Users {
+			variableValues, err := applyValueTraits(user, traits)
+			if err != nil {
+				if !trace.IsNotFound(err) {
+					log.WithError(err).Debugf("Skipping impersonate user %q.", user)
+				}
+				continue
+			}
+			outCond.Users = append(outCond.Users, variableValues...)
+		}
+		for _, role := range inCond.Roles {
+			variableValues, err := applyValueTraits(role, traits)
+			if err != nil {
+				if !trace.IsNotFound(err) {
+					log.WithError(err).Debugf("Skipping impersonate role %q.", role)
+				}
+				continue
+			}
+			outCond.Roles = append(outCond.Roles, variableValues...)
+		}
+		outCond.Users = utils.Deduplicate(outCond.Users)
+		outCond.Roles = utils.Deduplicate(outCond.Roles)
+		r.SetImpersonateConditions(condition, outCond)
```

Comment:
Don't you also need to copy Where to outCond? `outCond.Where = inCond.Where`. Otherwise seems like it might be lost here.